src/providers/idp/idp_auth_eval.c

```diff
@@ -0,0 +1,160 @@
+/*
+    SSSD
+
+    IdP Identity Backend Module - evalute authentication replies
+
+    Authors:
+        Sumit Bose <sbose@redhat.com>
+
+    Copyright (C) 2024 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <errno.h>
+#include <jansson.h>
+
+#include "util/util.h"
+#include "providers/idp/idp_auth.h"
+#include "krb5_plugin/idp/idp.h"
+#include "providers/krb5/krb5_utils.h"
+
+errno_t eval_device_auth_buf(struct idp_auth_ctx *idp_auth_ctx,
+                             struct pam_data *pd,
+                             uint8_t *buf, ssize_t buflen)
+{
+    uint8_t *user_reply;
+    uint8_t *end;
+    struct sss_idp_oauth2 *user_data = NULL;
+    int ret;
+    struct idp_open_req_data *open_req = NULL;
+
+    user_reply = memchr(buf, '\n', buflen);
+    if (user_reply == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Missing seperator in device auth reply.\n");
+        return EINVAL;
+    }
+
+    *user_reply = '\0';
+    user_reply++;
+
+    end = memchr(user_reply, '\n', buflen - (user_reply - 1 - buf));
```

Comment:
Why '-1'?

Reply:
I removed `-1`, this might be a leftover from an older version of the code.